### ISSUE TITLE
Build issues with latest hamlib and UHD installed in custom prefix

### ISSUE THREAD

**Issue by dkozel** [kb1vc/SoDaRadio#2]
I'm trying to build the latest version of SoDaRadio on Ubuntu 17.10 with the latest versions of UHD and hamlib installed from source. Additionally I prefer to install all software built from source into a custom prefix, `~/local`. I realize this is not likely a tested set of conditions.

I'm on Ubuntu 17.10 with 
UHD: https://github.com/EttusResearch/uhd/commit/4844f66dccaa71da102b02bba7b8caf8f84a932c
Hamlib: https://github.com/Hamlib/Hamlib/commit/7893b9f6d813ac53323db4efa9003380c3723a96
SoDaRadio code: 390a3c27

cmake command
```
[dkozel@er-dk-merlot build] cmake -DCMAKE_INSTALL_PREFIX=~/local -Wno-dev ..
-- The C compiler identification is GNU 7.2.0
-- The CXX compiler identification is GNU 7.2.0
-- Check for working C compiler: /usr/lib/ccache/cc
-- Check for working C compiler: /usr/lib/ccache/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
SoDaRadio_VERSION=[5.1.5]
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Found UHD: /home/dkozel/local/lib/libuhd.so  
Building qt gui
-- Found Qwt: /usr/lib/libqwt-qt5.so (found version "6.1.2") 
-- Checking for one of the modules 'hamlib'
-- Checking for one of the modules 'OPTIONAL;libgps'
FOUND GPSLIB
-- Found FFTW3F: /usr/lib/x86_64-linux-gnu/libfftw3f.so  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dkozel/src/SoDaRadio/build
```

Build error
```
....
/home/dkozel/src/SoDaRadio/qtgui/soda_hamlib_handler.cpp:32:10: fatal error: hamlib/rig.h: No such file or directory
 #include <hamlib/rig.h>
          ^~~~~~~~~~~~~~
compilation terminated.
qtgui/CMakeFiles/SoDaRadio.dir/build.make:392: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_handler.cpp.o' failed
make[2]: *** [qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_handler.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/dkozel/src/SoDaRadio/qtgui/soda_hamlib_server.cpp:36:10: fatal error: hamlib/rig.h: No such file or directory
 #include <hamlib/rig.h>
          ^~~~~~~~~~~~~~
compilation terminated.
qtgui/CMakeFiles/SoDaRadio.dir/build.make:379: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_server.cpp.o' failed
make[2]: *** [qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_server.cpp.o] Error 1
/home/dkozel/src/SoDaRadio/qtgui/mainwindow.cpp:33:10: fatal error: uhd/version.hpp: No such file or directory
 #include <uhd/version.hpp>
          ^~~~~~~~~~~~~~~~~
compilation terminated.
qtgui/CMakeFiles/SoDaRadio.dir/build.make:119: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/mainwindow.cpp.o' failed
make[2]: *** [qtgui/CMakeFiles/SoDaRadio.dir/mainwindow.cpp.o] Error 1
CMakeFiles/Makefile2:158: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/all' failed
make[1]: *** [qtgui/CMakeFiles/SoDaRadio.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

And the files do exist, so this is almost certainly a paths problem.
```
[dkozel@er-dk-merlot build] find ~/local/ -name version*
/home/dkozel/local/include/uhd/version.hpp
[dkozel@er-dk-merlot build] find ~/local/ -name rig.h
/home/dkozel/local/include/hamlib/rig.h
```

I'll continue debugging tomorrow evening, but wanted to file an issue with the info I have now incase this is something you've seen before. Thanks for all your work on this software!

**Comment by kb1vc** [comment]
Derek, 

Thanks for trying SoDaRadio.  (And thank you for including the git hash for the version that
you are trying to build....!)  

I think you are running into a problem with the cmake technique I'm using to find hamlib.  It 
finds the version installed in the system directories (or is supposed to) with the pkg_search_module
function in CMakeLists.txt.  

You might try something like this:  (in the build area)

cmake -DHAMLIB_LIBRARY_DIRS=~/.local/lib -DHAMLIB_INCLUDE_DIRS=~/.local/include ../

and see if that improves things.  

Do you want hamlib support (to connect to fldigi and wsjtx?)  If not, then I think you might
try commenting out the pkg_...(HAMLIB... line and see what happens. 

---------

It will be a day or two before I can work on this, I appreciate your patience.

73
matt


Sent from iCloud

On Mar 21, 2018, at 07:22 PM, Derek Kozel <notifications@github.com> wrote:

I'm trying to build the latest version of SoDaRadio on Ubuntu 17.10 with the latest versions of UHD and hamlib installed from source. Additionally I prefer to install all software built from source into a custom prefix, ~/local. I realize this is not likely a tested set of conditions.

I'm on Ubuntu 17.10 with
UHD: EttusResearch/uhd@4844f66
Hamlib: Hamlib/Hamlib@7893b9f
SoDaRadio code: 390a3c2

cmake command

[dkozel@er-dk-merlot build] cmake -DCMAKE_INSTALL_PREFIX=~/local -Wno-dev ..
-- The C compiler identification is GNU 7.2.0
-- The CXX compiler identification is GNU 7.2.0
-- Check for working C compiler: /usr/lib/ccache/cc
-- Check for working C compiler: /usr/lib/ccache/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
SoDaRadio_VERSION=[5.1.5]
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Found UHD: /home/dkozel/local/lib/libuhd.so  
Building qt gui
-- Found Qwt: /usr/lib/libqwt-qt5.so (found version "6.1.2") 
-- Checking for one of the modules 'hamlib'
-- Checking for one of the modules 'OPTIONAL;libgps'
FOUND GPSLIB
-- Found FFTW3F: /usr/lib/x86_64-linux-gnu/libfftw3f.so  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dkozel/src/SoDaRadio/build
Build error

....
/home/dkozel/src/SoDaRadio/qtgui/soda_hamlib_handler.cpp:32:10: fatal error: hamlib/rig.h: No such file or directory
 #include <hamlib/rig.h>
          ^~~~~~~~~~~~~~
compilation terminated.
qtgui/CMakeFiles/SoDaRadio.dir/build.make:392: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_handler.cpp.o' failed
make[2]: *** [qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_handler.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/dkozel/src/SoDaRadio/qtgui/soda_hamlib_server.cpp:36:10: fatal error: hamlib/rig.h: No such file or directory
 #include <hamlib/rig.h>
          ^~~~~~~~~~~~~~
compilation terminated.
qtgui/CMakeFiles/SoDaRadio.dir/build.make:379: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_server.cpp.o' failed
make[2]: *** [qtgui/CMakeFiles/SoDaRadio.dir/soda_hamlib_server.cpp.o] Error 1
/home/dkozel/src/SoDaRadio/qtgui/mainwindow.cpp:33:10: fatal error: uhd/version.hpp: No such file or directory
 #include <uhd/version.hpp>
          ^~~~~~~~~~~~~~~~~
compilation terminated.
qtgui/CMakeFiles/SoDaRadio.dir/build.make:119: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/mainwindow.cpp.o' failed
make[2]: *** [qtgui/CMakeFiles/SoDaRadio.dir/mainwindow.cpp.o] Error 1
CMakeFiles/Makefile2:158: recipe for target 'qtgui/CMakeFiles/SoDaRadio.dir/all' failed
make[1]: *** [qtgui/CMakeFiles/SoDaRadio.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
And the files do exist, so this is almost certainly a paths problem.

[dkozel@er-dk-merlot build] find ~/local/ -name version*
/home/dkozel/local/include/uhd/version.hpp
[dkozel@er-dk-merlot build] find ~/local/ -name rig.h
/home/dkozel/local/include/hamlib/rig.h
I'll continue debugging tomorrow evening, but wanted to file an issue with the info I have now incase this is something you've seen before. Thanks for all your work on this software!

**Comment by dkozel** [comment]
Thanks for the tip. I bet that would work. Yes, I want to use wsjtx. I'll give this a shot as soon as I can and let you know if that works. Could the include of the UHD version.hpp have the same issue? The FindUHD.cmake routine should be working and supplying the correct include paths. I'll try looking at the generated Makefile when I get a chance.

**Comment by kb1vc** [comment]
Derek, 

The general solution is 
`cmake -DPKG_CONFIG_USE_CMAKE_PREFIX_PATH=1 -DCMAKE_PREFIX_PATH=${HOME}/.local ../`

You'll need to wipe the directory where you're running cmake, as otherwise it will use the cached path resolutions.   By adding the CMAKE_PREFIX_PATH the various pkg config scripts should look in your ~/.local tree first.  

Please let me know if this fixes the problem.  My recollection was that there was a bit of a problem with wsjtx having to do with their using a "patched" version of hamlib.  I think it meant that wsjtx needed the library and that SoDaRadio could link against the "stock" version of the library.  (But I may be wrong...)

**Comment by kb1vc** [comment]
Man what a mess.  The CMake files worked under Fedora and didn't work under Ubuntu 16 or 17. 

The PKG_CONFIG_USE_CMAKE_PREFIX_PATH and CMAKE_PREFIX_PATH stuff addresses the presenting condition : the need to select alternate installation directories.  But even if that worked, the Qt build would fail.  This is fixed in master 775a671cd
